File: quant1x/contrib/data/tdx/level1/std/xdxr_info.go
package std

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"math"

	"github.com/quant1x/quant1x/quant1x/contrib/data/tdx/tdxproto"
	"github.com/quant1x/quant1x/quant1x/data"
)

// XdxrCategory 除权除息类型枚举
type XdxrCategory int

const (
	ExDividend                     XdxrCategory = 1  // 除权除息
	BonusSharesListing             XdxrCategory = 2  // 送股上市(无偿)
	RestrictedSharesListing        XdxrCategory = 3  // 非流通股上市(受限股解禁)
	UnspecifiedCapitalAdjustment   XdxrCategory = 4  // 未知股本变动
	GeneralCapitalAdjustment       XdxrCategory = 5  // 股本变化(保留, 但慎用)
	NewShareIssuance               XdxrCategory = 6  // 增发新股
	ShareRepurchase                XdxrCategory = 7  // 股份回购
	NewSharesListing               XdxrCategory = 8  // 增发新股上市
	TransferredRightsSharesListing XdxrCategory = 9  // 转配股上市(中国特有)
	ConvertibleBondListing         XdxrCategory = 10 // 可转债上市
	StockSplitOrReverseSplit       XdxrCategory = 11 // 拆股或合股
	RestrictedSharesConsolidation  XdxrCategory = 12 // 非流通股缩股
	IssueCallWarrants              XdxrCategory = 13 // 送认购权证
	IssuePutWarrants               XdxrCategory = 14 // 送认沽权证
)

// ToString 将枚举值转换为描述文本
func (c XdxrCategory) ToString() string {
	switch c {
	case ExDividend:
		return "除权除息"
	case BonusSharesListing:
		return "送配股上市"
	case RestrictedSharesListing:
		return "非流通股上市"
	case UnspecifiedCapitalAdjustment:
		return "未知股本变动"
	case GeneralCapitalAdjustment:
		return "股本变化"
	case NewShareIssuance:
		return "增发新股"
	case ShareRepurchase:
		return "股份回购"
	case NewSharesListing:
		return "增发新股上市"
	case TransferredRightsSharesListing:
		return "转配股上市"
	case ConvertibleBondListing:
		return "可转债上市"
	case StockSplitOrReverseSplit:
		return "扩缩股"
	case RestrictedSharesConsolidation:
		return "非流通股缩股"
	case IssueCallWarrants:
		return "送认购权证"
	case IssuePutWarrants:
		return "送认沽权证"
	default:
		return fmt.Sprintf("Unknown(%d)", int(c))
	}
}

// XdxrInfo represents a parsed XDXR event returned by the server.
type XdxrInfo struct {
	Date          string  // 日期 YYYY-MM-DD格式
	Category      int     // 类型编号
	Name          string  // 类型名称
	FenHong       float64 // 分红(元)
	PeiGuJia      float64 // 配股价(元)
	SongZhuanGu   float64 // 送转股(股)
	PeiGu         float64 // 配股(股)
	SuoGu         float64 // 缩股(股)
	QianLiuTong   float64 // 除权前流通股(万股)
	HouLiuTong    float64 // 除权后流通股(万股)
	QianZongGuBen float64 // 除权前总股本(万股)
	HouZongGuBen  float64 // 除权后总股本(万股)
	FenShu        float64 // 权证份数
	XingQuanJia   float64 // 行权价格(元)
}

// IsAdjust 是否进行除权除息调整
func (x XdxrInfo) IsAdjust() bool {
	count := x.FenHong     // 分红
	count += x.PeiGu       // 配股
	count += x.SongZhuanGu // 送转股
	count += x.SuoGu       // 缩股
	count += x.FenShu      // 行权
	return count > 0.0
}

// AdjustFactor 计算调整因子m和a
func (x XdxrInfo) AdjustFactor() (float64, float64) {
	var m, a float64

	A := x.ComputeMonetaryAdjustment()
	B := x.ComputeShareAdjustmentRatio()

	if math.Abs(1.0+B) > 1e-10 {
		m = 1.0 / (1.0 + B)
		if m < 0 {
			m = 1.0
		}
		a = A * m
	} else {
		m = 1.0
		a = 0.0
	}

	return m, a
}

func (x XdxrInfo) ComputeMonetaryAdjustment() float64 {
	return (x.PeiGu*x.PeiGuJia - x.FenHong + x.FenShu*x.XingQuanJia) / 10.0
}

func (x XdxrInfo) ComputeShareAdjustmentRatio() float64 {
	return (x.SongZhuanGu + x.PeiGu - x.SuoGu + x.FenShu) / 10.0
}

func (x XdxrInfo) IsCapitalChange() bool {
	switch x.Category {
	case int(ExDividend),
		int(StockSplitOrReverseSplit),
		int(RestrictedSharesConsolidation),
		int(IssueCallWarrants),
		int(IssuePutWarrants):
		return false
	default:
		if x.HouLiuTong > 0 && x.HouZongGuBen > 0 {
			return true
		}
	}
	return false
}

// Adjust 生成复权计算函数
func (x XdxrInfo) Adjust() func(float64) float64 {
	songZhuangu := x.SongZhuanGu
	peiGu := x.PeiGu
	suoGu := x.SuoGu
	xdxrGuShu := (songZhuangu + peiGu - suoGu) / 10
	fenHong := x.FenHong
	peiGuJia := x.PeiGuJia
	xdxrFenHong := (peiGuJia*peiGu - fenHong) / 10

	return func(p float64) float64 {
		return (p + xdxrFenHong) / (1 + xdxrGuShu)
	}
}

func (x XdxrInfo) String() string {
	return fmt.Sprintf("Date: %s Category: %d Name: %s FenHong: %f PeiGuJia: %f SongZhuanGu: %f PeiGu: %f SuoGu: %f QianLiuTong: %f HouLiuTong: %f QianZongGuBen: %f HouZongGuBen: %f FenShu: %f XingQuanJia: %f",
		x.Date, x.Category, x.Name, x.FenHong, x.PeiGuJia, x.SongZhuanGu, x.PeiGu, x.SuoGu, x.QianLiuTong, x.HouLiuTong, x.QianZongGuBen, x.HouZongGuBen, x.FenShu, x.XingQuanJia)
}

// XdxrInfoContext 对齐 C++/Rust/Python XdxrInfoContext, 合并请求和响应.
type XdxrInfoContext struct {
	tdxproto.FrameBase
	Instrument data.InstrumentInfo
	Count      uint16
	List       []XdxrInfo
}

// NewXdxrInfoContext 构造除权除息请求, 对齐 C++/Rust.
func NewXdxrInfoContext(instrument data.InstrumentInfo) *XdxrInfoContext {
	return &XdxrInfoContext{
		FrameBase:  tdxproto.NewFrameBase(tdxproto.StdCommandXdxrInfo, tdxproto.FlagUncompressed, tdxproto.PacketTypeRequest),
		Instrument: instrument,
	}
}

// SerializeRequestBody 序列化请求体, 对齐 C++/Rust/Python.
func (x *XdxrInfoContext) SerializeRequestBody() []byte {
	payload := &bytes.Buffer{}
	padding := []byte{0x01, 0x00}
	payload.Write(padding)
	market := uint8(tdxproto.ExchangeToMarketId(x.Instrument.Exchange))
	payload.WriteByte(market)
	payload.Write([]byte(x.Instrument.Ticker)[:6])
	return payload.Bytes()
}

// DeserializeResponseBody 解析除权除息响应体, 对齐 C++/Rust/Python.
func (x *XdxrInfoContext) DeserializeResponseBody(body []byte) error {
	reader := bytes.NewReader(body)
	// skip 9 bytes as in C++ (Unknown header)
	if _, err := reader.Seek(9, io.SeekStart); err != nil {
		return err
	}
	if err := binary.Read(reader, binary.LittleEndian, &x.Count); err != nil {
		return err
	}
	x.List = make([]XdxrInfo, 0, int(x.Count))
	for i := 0; i < int(x.Count); i++ {
		var market uint8
		if err := binary.Read(reader, binary.LittleEndian, &market); err != nil {
			return err
		}
		codeBuf := make([]byte, 6)
		if _, err := io.ReadFull(reader, codeBuf); err != nil {
			return err
		}
		// unknown byte
		var _unk uint8
		if err := binary.Read(reader, binary.LittleEndian, &_unk); err != nil {
			return err
		}
		var dateRaw uint32
		if err := binary.Read(reader, binary.LittleEndian, &dateRaw); err != nil {
			return err
		}
		var category uint8
		if err := binary.Read(reader, binary.LittleEndian, &category); err != nil {
			return err
		}
		data := make([]byte, 16)
		if _, err := io.ReadFull(reader, data); err != nil {
			return err
		}

		y, m, d, _, _ := tdxproto.GetDatetimeFromUint32(9, dateRaw, 0)
		xi := XdxrInfo{Date: fmt.Sprintf("%04d-%02d-%02d", y, m, d), Category: int(category), Name: XdxrCategory(category).ToString()}

		// parse data per category similar to C++ logic
		db := bytes.NewReader(data)
		switch category {
		case 1: // 除权除息
			var f32v float32
			_ = binary.Read(db, binary.LittleEndian, &f32v)
			xi.FenHong = float64(f32v)
			_ = binary.Read(db, binary.LittleEndian, &f32v)
			xi.PeiGuJia = float64(f32v)
			_ = binary.Read(db, binary.LittleEndian, &f32v)
			xi.SongZhuanGu = float64(f32v)
			_ = binary.Read(db, binary.LittleEndian, &f32v)
			xi.PeiGu = float64(f32v)
		case 11, 12:
			if _, err := db.Seek(8, io.SeekStart); err == nil {
				var f32v float32
				_ = binary.Read(db, binary.LittleEndian, &f32v)
				xi.SuoGu = float64(f32v)
			}
		case 13, 14:
			var f32v float32
			_ = binary.Read(db, binary.LittleEndian, &f32v)
			xi.XingQuanJia = float64(f32v)
			if _, err := db.Seek(8, io.SeekCurrent); err == nil {
				var f32v2 float32
				_ = binary.Read(db, binary.LittleEndian, &f32v2)
				xi.FenShu = float64(f32v2)
			}
		default:
			var v uint32
			_ = binary.Read(db, binary.LittleEndian, &v)
			xi.QianLiuTong = tdxproto.IntegerToFloat64(v)
			_ = binary.Read(db, binary.LittleEndian, &v)
			xi.QianZongGuBen = tdxproto.IntegerToFloat64(v)
			_ = binary.Read(db, binary.LittleEndian, &v)
			xi.HouLiuTong = tdxproto.IntegerToFloat64(v)
			_ = binary.Read(db, binary.LittleEndian, &v)
			xi.HouZongGuBen = tdxproto.IntegerToFloat64(v)
		}

		x.List = append(x.List, xi)
	}
	return nil
}

func (x *XdxrInfoContext) String() string {
	return fmt.Sprintf("XdxrInfoContext{%s, Count:%d}", x.Instrument.Symbol(), x.Count)
}

File: quant1x/contrib/data/tdx/level1/std/xdxr_info.rs
// -*- coding: utf-8 -*-
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// xdxr — 除权除息信息 (STD_XDXR_INFO, 0x000f)
// 对应 Python level1/std/xdxr.py
// 对应 C++   level1/xdxr_info.h

use crate::std::BinaryStream;
use crate::data::schema::{XdxrInfo, XdxrEntry};
use crate::data::meta::Exchange;

use super::super::super::command::*;
use super::super::super::helpers::{get_datetime_from_u32, int_to_float64, exchange_to_market, market_to_exchange};
use super::super::super::protocol::{BaseFrame, RequestHeader, ResponseHeader};

// ============================================================
// XdxrInfoContext — 单只证券除权除息查询
// 对应 Python class Xdxr
// ============================================================

/// 单只证券除权除息请求/响应
#[derive(Debug, Clone)]
pub struct XdxrInfoContext {
    req_header: RequestHeader,
    resp_header: ResponseHeader,
    /// TDX 市场编号 (0:深圳, 1:上海)
    pub market: u8,
    /// 证券代码 (6字节 ASCII)
    pub ticker: String,
    /// 返回的记录条数
    pub count: u16,
    /// 除权除息记录列表
    pub list: Vec<XdxrInfo>,
}

impl XdxrInfoContext {
    pub fn new(exchange: Exchange, ticker: &str) -> Self {
        let market_id = exchange_to_market(exchange.code())
            .unwrap_or(0) as u8;

        Self {
            req_header: RequestHeader::new(STD_XDXR_INFO, FLAG_UNCOMPRESSED),
            resp_header: ResponseHeader::new(),
            market: market_id,
            ticker: ticker.to_string(),
            count: 0,
            list: Vec::new(),
        }
    }
}

impl BaseFrame for XdxrInfoContext {
    fn request_header(&self) -> &RequestHeader { &self.req_header }
    fn request_header_mut(&mut self) -> &mut RequestHeader { &mut self.req_header }
    fn response_header(&self) -> &ResponseHeader { &self.resp_header }
    fn response_header_mut(&mut self) -> &mut ResponseHeader { &mut self.resp_header }

    fn serialize_request_body(&mut self) -> Vec<u8> {
        let mut bs = BinaryStream::new();
        // padding: bytes.fromhex('0100') = [0x01, 0x00]
        bs.push_u8(0x01);
        bs.push_u8(0x00);
        // market: u8
        bs.push_u8(self.market);
        // code: 6 bytes ASCII, pad with \0 if shorter
        let mut code_bytes = [0u8; 6];
        let ticker_bytes = self.ticker.as_bytes();
        let len = ticker_bytes.len().min(6);
        code_bytes[..len].copy_from_slice(&ticker_bytes[..len]);
        bs.push_byte_array(&code_bytes);
        bs.data().clone()
    }

    fn deserialize_response_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        self.list.clear();
        if data.len() < 11 {
            // 9 bytes unknown + 2 bytes count = 11 minimum
            return Ok(());
        }

        let mut bs = BinaryStream::from_vec(data.to_vec());
        let data_len = data.len();

        // skip 9 bytes unknown header
        bs.skip(9);

        // count: u16
        let remaining = data_len.saturating_sub(bs.position());
        if remaining < 2 {
            return Ok(());
        }
        self.count = bs.get_u16()?;

        for _ in 0..self.count {
            let remaining = data_len.saturating_sub(bs.position());
            if remaining < 29 {
                log::warn!("Insufficient data when parsing XDXR_INFO payload");
                break;
            }

            // market: u8 (skip)
            bs.skip(1);

            // code: 6 bytes ASCII (skip)
            bs.skip(6);

            // unknown: u8 (skip)
            bs.skip(1);

            // date: u32
            let date_int = bs.get_u32()?;

            // category: u8
            let category = bs.get_u8()? as i32;

            // record_data: 16 bytes
            let mut record_data = [0u8; 16];
            bs.get_byte_array(&mut record_data)?;

            // parse date
            let (year, month, day, _, _) = get_datetime_from_u32(9, date_int, 0);

            let mut info = XdxrInfo {
                date: format!("{:04}-{:02}-{:02}", year, month, day),
                category,
                name: crate::data::schema::XdxrCategory::to_string(category),
                ..Default::default()
            };

            // parse record_data based on category
            let mut tmp = BinaryStream::from_vec(record_data.to_vec());
            match category {
                1 => {
                    // 除权除息
                    info.fen_hong = tmp.get_f32()? as f64;
                    info.pei_gu_jia = tmp.get_f32()? as f64;
                    info.song_zhuan_gu = tmp.get_f32()? as f64;
                    info.pei_gu = tmp.get_f32()? as f64;
                }
                11 | 12 => {
                    // 扩缩股 / 非流通股缩股
                    tmp.skip(8);
                    info.suo_gu = tmp.get_f32()? as f64;
                }
                13 | 14 => {
                    // 送认购权证 / 送认沽权证
                    info.xing_quan_jia = tmp.get_f32()? as f64;
                    tmp.skip(8);
                    info.fen_shu = tmp.get_f32()? as f64;
                }
                _ => {
                    // 其他类型: 股本变动
                    let v1 = tmp.get_u32()?;
                    info.qian_liu_tong = Self::get_v(v1);
                    let v2 = tmp.get_u32()?;
                    info.qian_zong_gu_ben = Self::get_v(v2);
                    let v3 = tmp.get_u32()?;
                    info.hou_liu_tong = Self::get_v(v3);
                    let v4 = tmp.get_u32()?;
                    info.hou_zong_gu_ben = Self::get_v(v4);
                }
            }

            self.list.push(info);
        }

        log::debug!("xdxr fetched market={} ticker={} count={} parsed={}",
                    self.market, self.ticker, self.count, self.list.len());

        Ok(())
    }
}

impl XdxrInfoContext {
    /// 将 u32 整数转换为 f64 浮点数(与 level1 协议中使用的转换一致)
    fn get_v(v: u32) -> f64 {
        if v == 0 {
            return 0.0;
        }
        int_to_float64(v)
    }
}

// ============================================================
// XdxrBatchRequest — 批量除权除息查询
// 对应 Python class XdxrBatch
// ============================================================

/// 批量除权除息请求/响应
#[derive(Debug, Clone)]
pub struct XdxrBatchRequest {
    req_header: RequestHeader,
    resp_header: ResponseHeader,
    /// 证券列表 (exchange, ticker)
    pub instruments: Vec<(Exchange, String)>,
    /// 返回的条目数量
    pub count: u16,
    /// 批量除权除息条目列表
    pub list: Vec<XdxrEntry>,
}

impl XdxrBatchRequest {
    pub fn new(instruments: Vec<(Exchange, String)>) -> Self {
        Self {
            req_header: RequestHeader::new(STD_XDXR_INFO, FLAG_UNCOMPRESSED),
            resp_header: ResponseHeader::new(),
            instruments,
            count: 0,
            list: Vec::new(),
        }
    }
}

impl BaseFrame for XdxrBatchRequest {
    fn request_header(&self) -> &RequestHeader { &self.req_header }
    fn request_header_mut(&mut self) -> &mut RequestHeader { &mut self.req_header }
    fn response_header(&self) -> &ResponseHeader { &self.resp_header }
    fn response_header_mut(&mut self) -> &mut ResponseHeader { &mut self.resp_header }

    fn serialize_request_body(&mut self) -> Vec<u8> {
        let mut bs = BinaryStream::new();
        let inst_count = self.instruments.len() as u16;
        bs.push_u16(inst_count);

        for (exchange, ticker) in &self.instruments {
            let market_id = exchange_to_market(exchange.code()).unwrap_or(0) as u8;
            bs.push_u8(market_id);

            // code: 6 bytes ASCII, pad with \0 if shorter
            let mut code_bytes = [0u8; 6];
            let ticker_bytes = ticker.as_bytes();
            let len = ticker_bytes.len().min(6);
            code_bytes[..len].copy_from_slice(&ticker_bytes[..len]);
            bs.push_byte_array(&code_bytes);
        }

        bs.data().clone()
    }

    fn deserialize_response_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        self.list.clear();
        if data.len() < 2 {
            return Ok(());
        }

        let mut bs = BinaryStream::from_vec(data.to_vec());
        let data_len = data.len();

        self.count = bs.get_u16()?;

        for _ in 0..self.count {
            let remaining = data_len.saturating_sub(bs.position());
            if remaining < 9 {
                log::warn!("Insufficient data when parsing XDXR_BATCH header");
                break;
            }

            // market_id: u8
            let market_id = bs.get_u8()? as i32;

            // ticker: 6 bytes ASCII
            let mut code_bytes = [0u8; 6];
            bs.get_byte_array(&mut code_bytes)?;
            let ticker = String::from_utf8_lossy(&code_bytes)
                .trim_end_matches('\0')
                .trim()
                .to_string();

            // sub_count: u16
            let sub_count = bs.get_u16()? as usize;

            let exchange_str = market_to_exchange(market_id).unwrap_or("SSE");
            let exchange = Exchange::parse(exchange_str)
                .unwrap_or(Exchange::SSE);

            let mut entry = XdxrEntry {
                exchange,
                ticker: ticker.clone(),
                count: sub_count as i32,
                list: Vec::with_capacity(sub_count),
            };

            for _ in 0..sub_count {
                let remaining = data_len.saturating_sub(bs.position());
                if remaining < 29 {
                    log::warn!("Insufficient data when parsing XDXR_BATCH record");
                    break;
                }

                // market (skip)
                bs.skip(1);
                // code (skip)
                bs.skip(6);
                // unknown (skip)
                bs.skip(1);

                // date: u32
                let date_int = bs.get_u32()?;

                // category: u8
                let category = bs.get_u8()? as i32;

                // record_data: 16 bytes
                let mut record_data = [0u8; 16];
                bs.get_byte_array(&mut record_data)?;

                // parse date
                let (year, month, day, _, _) = get_datetime_from_u32(9, date_int, 0);

                let mut info = XdxrInfo {
                    date: format!("{:04}-{:02}-{:02}", year, month, day),
                    category,
                    name: crate::data::schema::XdxrCategory::to_string(category),
                    ..Default::default()
                };

                // parse record_data based on category
                let mut tmp = BinaryStream::from_vec(record_data.to_vec());
                match category {
                    1 => {
                        info.fen_hong = tmp.get_f32()? as f64;
                        info.pei_gu_jia = tmp.get_f32()? as f64;
                        info.song_zhuan_gu = tmp.get_f32()? as f64;
                        info.pei_gu = tmp.get_f32()? as f64;
                    }
                    11 | 12 => {
                        tmp.skip(8);
                        info.suo_gu = tmp.get_f32()? as f64;
                    }
                    13 | 14 => {
                        info.xing_quan_jia = tmp.get_f32()? as f64;
                        tmp.skip(8);
                        info.fen_shu = tmp.get_f32()? as f64;
                    }
                    _ => {
                        let v1 = tmp.get_u32()?;
                        info.qian_liu_tong = XdxrInfoContext::get_v(v1);
                        let v2 = tmp.get_u32()?;
                        info.qian_zong_gu_ben = XdxrInfoContext::get_v(v2);
                        let v3 = tmp.get_u32()?;
                        info.hou_liu_tong = XdxrInfoContext::get_v(v3);
                        let v4 = tmp.get_u32()?;
                        info.hou_zong_gu_ben = XdxrInfoContext::get_v(v4);
                    }
                }

                entry.list.push(info);
            }

            self.list.push(entry);
        }

        log::debug!("xdxr_batch fetched count={} parsed={}",
                    self.count, self.list.len());

        Ok(())
    }
}

// ============================================================
// fetch_xdxr — 便捷函数
// ============================================================

/// 获取单只证券的除权除息信息
pub fn fetch_xdxr(exchange: Exchange, ticker: &str) -> Option<XdxrInfoContext> {
    match super::super::super::client::get_std_conn() {
        Ok(mut conn) => {
            let mut msg = XdxrInfoContext::new(exchange, ticker);
            match super::super::super::protocol::transact_message_sync(conn.stream(), &mut msg) {
                Ok(()) => Some(msg),
                Err(e) => {
                    log::error!("level1 xdxr process error for {} {}: {}", exchange.code(), ticker, e);
                    None
                }
            }
        }
        Err(e) => {
            log::error!("failed to acquire level1 client for xdxr {} {}: {}", exchange.code(), ticker, e);
            None
        }
    }
}

/// 批量获取多只证券的除权除息信息
pub fn fetch_xdxr_batch(instruments: Vec<(Exchange, String)>) -> Option<XdxrBatchRequest> {
    if instruments.is_empty() {
        return None;
    }
    match super::super::super::client::get_std_conn() {
        Ok(mut conn) => {
            let mut msg = XdxrBatchRequest::new(instruments);
            match super::super::super::protocol::transact_message_sync(conn.stream(), &mut msg) {
                Ok(()) => Some(msg),
                Err(e) => {
                    log::error!("level1 xdxr_batch process error: {}", e);
                    None
                }
            }
        }
        Err(e) => {
            log::error!("failed to acquire level1 client for xdxr_batch: {}", e);
            None
        }
    }
}
